POD Specialization constants design update (#2669)
intel/llvm: sycl/doc/SpecializationConstants.md

Before JIT-ing a program, the runtime "flushes" the spec constants: it iterates
through the value map and invokes the

> ```c++

Reviewer: Right, Sphinx doesn't understand this syntax.

> ```c++

Author: Does it completely breaks sphinx rendering?

Looking at the docs, I'm a bit confused: [Sphinx](https://www.sphinx-doc.org/en/master/usage/markdown.html) uses [recommonmark](https://recommonmark.readthedocs.io/en/latest/index.html), which in turn has [AutoStrictify](https://recommonmark.readthedocs.io/en/latest/auto_structify.html#codeblock-extensions) component for converting Markdown into RST and there is an example of codeblock with language in the docs. Are we using different version, or it is a configuration issue, or bug in the documentation?

> ```c++

Reviewer: > Does it completely break sphinx rendering?

It does. I was also surprised by the fact that it does not understand this syntax. Maybe, it's a bug that'll be fixed in the future. But at the moment this doesn't work.

> ```c++

Author: Ok, reverted in 69bdfdeaabfd95ee589cde9c9f3b2a9c64c6644a

Say, the POD type is

```c++
struct A {
  int x;
  float y;
};

struct POD {
  A a[2];

>   cl::sycl::vec<int, 2> b;

Author: This example might actually be confusing, because `cl::sycl::vec` is not a POD-type in our implementation, because it is not trivially copyable due to the fact that there is non-trivial copy constructor (I don't know why it is coded in this way, probably @romanovvlad can say something about this).

I guess that from user point of view it would be nice to have possibility to use vectors as specialization constants, so, what should I do here: do I put a vector into the example or do I leave it as-is?

       %i32 = OpSpecConstant %int.type 0                       ; 1st member with default value
     %float = OpSpecConstant %float.type 0.0                   ; 2nd member with default value
    %struct = OpSpecConstantComposite %struct.type %i32 %float ; Composite doens't need IDs or default value
         %1 = OpTypeFunction %struct.type

       %get = OpFunction %struct.type None %1
         %2 = OpLabel
              OpReturnValue %struct
              OpFunctionEnd
```

##### The post-link tool changes

For composite specialization constants the post link tool will additionally
generate linearized list of \<leaf spec ID,offset,size\> tuples (descriptors),
where each tuple describes a leaf field, and store it together with the
existing meta-information associated with the specialization constants and
passed to the runtime. Also, for a composite specialization constant there is
no ID map entry within the meta information, and the composite constant is
referenced by its symbolic ID. For example:

```
MyConst_mangled [10,0,4],[11,4,4],[12,8,4],[13,12,4],[14,16,4]
```

This tuple is needed, because at SYCL runtime level, composite constants are set

> by user as a single entity and we have to break it down to its members and set a

Reviewer 2: ```suggestion
by user as a byte array and we have to break it down to the leaf members of the composite and set a
```

Author: Applied in 8f323a1bcaae66ee26f4918e028f0955a0c104fa

> value for each leaf as for a separate specialization constant. It contains the

Reviewer 2: ```suggestion
value for each leaf as for a separate scalar specialization constant. Each tuple contains the
```

Author: Applied in 8f323a1bcaae66ee26f4918e028f0955a0c104fa

following data:
- ID of composite constants leaf, i.e. ID of scalar spec constant

> - Offset from the beginning of composite, which points to the location of a

Reviewer 2: I guess the offset also points to the location of the specialization value in the byte array provided by the user. Should we mention it here?

Author: I've updated this item in 8f323a1bcaae66ee26f4918e028f0955a0c104fa

##### The SpecConstant pass changes

The SpecConstants pass in the post-link will have the following IR as input
(`sret` conversion is omitted for clarity):

```
%struct.POD = type { [2 x %struct.A], <2 x i32> }
%struct.A = type { i32, float }

%spec_const = call %struct.POD __sycl_getCompositeSpecConstantValue<POD type mangling> ("MyConst_mangled")
```

`__sycl_getCompositeSpecConstantValue` is a new "intrinsic" (in addition to

> `__sycl_getSpecConstantValue`) recognized by `SpecConstants` pass, which creates

Reviewer 3: ```suggestion
`__sycl_getSpecConstantValue`) recognized by the `SpecConstants` pass, which creates
```

Author: Fixed in 6e248d5

a value of a composite (of non-primitive type) specialization constant.
It does not need a default value, because its default value consists of default
values of its leaf specialization constants (see below).

> `__sycl_getCompositeSpecConstantValue` will be replaced with a set of

Reviewer 3: please return the note about 'spec constant enumeration (symbolic -\> int ID translation' - this is essential job done by the pass

Author: Added such a note in 6e248d5

If any composite member is another composite, then it will be also represented
by number of `__spirv_SpecConstant` plus one `__spirv_SpecConstantComposite`:

```
%gold_POD_A0_x = call i32 __spirv_SpecConstant(i32 10, i32 0)
%gold_POD_A0_y = call float __spirv_SpecConstant(i32 11, float 0)

%gold_POD_A0 = call %struct.A __spirv_SpecConstantComposite(i32 %gold_POD_A0_x, float %gold_POD_A0_y)

%gold_POD_A1_x = call i32 __spirv_SpecConstant(i32 12, i32 0)
%gold_POD_A1_y = call float __spirv_SpecConstant(i32 13, float 0)

%gold_POD_A1 = call %struct.A __spirv_SpecConstantComposite(i32 %gold_POD_A1_x, float %gold_POD_A1_y)

%gold_POD_A = call [2 x %struct.A] __spirv_SpecConstantComposite(%struct.A %gold_POD_A0, %struct.A %gold_POD_A1)

%gold_POD_b0 = call i32 __spirv_SpecConstant(i32 14, i32 0)
%gold_POD_b1 = call i32 __spirv_SpecConstant(i32 15, i32 0)
%gold_POD_b = call <2 x i32> __spirv_SpecConstant(i32 %gold_POD_b0, i32 %gold_POD_b1)

%gold = call %struct.POD __spirv_SpecConstantComposite([2 x %struct.A] %gold_POD_A, <2 x i32> %gold_POD_b)

```

Spec ID for the composite spec constant is not needed, as runtime will never use
it - it will use IDs of the leaves instead.

> Yet, the SPIR-V specification does not allow `SpecID` decoration for composite

Reviewer 3: ```suggestion
Besides, the SPIR-V specification does not allow `SpecID` decoration for composite
```

Author: Fixed in 6e248d5

spec constants, because its defined by its members instead.

> `__spirv_SpecConstantComposite` is a new "intrinsic", which represents composite

Reviewer 3: ```suggestion
`__spirv_SpecConstantComposite` is a new SPIR-V intrinsic, which represents composite
```

Author: Fixed in 6e248d5

> ##### The SpecConstant pass changes

Reviewer 2: Why "changes"?
IMO the document should describe the design(probably with some implementation details) like it is _already done_, not like a _proposal_ to change the design or implementation.

Author: I guess it intended to meant "changes compared to scalars handling", but I agree with you, removed "changes" in 879566ac6f71c4a82d1bf2eb7933df584982b90c

The SpecConstants pass in the post-link will have the following IR as input
(`sret` conversion is omitted for clarity):

```
%struct.POD = type { [2 x %struct.A], <2 x i32> }
%struct.A = type { i32, float }

%spec_const = call %struct.POD __sycl_getCompositeSpecConstantValue<POD type mangling> ("MyConst_mangled")
```

`__sycl_getCompositeSpecConstantValue` is a new "intrinsic" (in addition to

> `__sycl_getSpecConstantValue`) recognized by the `SpecConstants` pass, which

Author: @kbobrovs, what do you think about renaming existing `__sycl_getSpecConstantValue` into `__sycl_getScalarSpecConstantValue` to unify names and add more clarity?

Reviewer 3: @AlexeySachkov, I don't mind

%gold_POD_A1_x = call i32 __spirv_SpecConstant(i32 12, i32 0)
%gold_POD_A1_y = call float __spirv_SpecConstant(i32 13, float 0)

%gold_POD_A1 = call %struct.A __spirv_SpecConstantComposite(i32 %gold_POD_A1_x, float %gold_POD_A1_y)

%gold_POD_A = call [2 x %struct.A] __spirv_SpecConstantComposite(%struct.A %gold_POD_A0, %struct.A %gold_POD_A1)

%gold_POD_b0 = call i32 __spirv_SpecConstant(i32 14, i32 0)
%gold_POD_b1 = call i32 __spirv_SpecConstant(i32 15, i32 0)
%gold_POD_b = call <2 x i32> __spirv_SpecConstant(i32 %gold_POD_b0, i32 %gold_POD_b1)

%gold = call %struct.POD __spirv_SpecConstantComposite([2 x %struct.A] %gold_POD_A, <2 x i32> %gold_POD_b)

```

Spec ID for the composite spec constant is not needed, as runtime will never use
it - it will use IDs of the leaves instead, which are being assigned by the
`SpecConstants` pass during replacement of SYCL intrinsics with SPIR-V
intrinsics.
Besides, the SPIR-V specification does not allow `SpecID` decoration for
composite spec constants, because its defined by its members instead.

`__spirv_SpecConstantComposite` is a new SPIR-V intrinsic, which represents

> composite specialization constant. Its arguments are LLVM IR valures

Reviewer 2: ```suggestion
composite specialization constant. Its arguments are LLVM IR values
```

Author: Fixed in 879566ac6f71c4a82d1bf2eb7933df584982b90c

> corresponding to elements of composite type of the constant.

Reviewer 2: ```suggestion
corresponding to elements of the constant composite .
```

Author: Fixed in 879566ac6f71c4a82d1bf2eb7933df584982b90c

##### LLVM -> SPIR-V translation

Given the `__spirv_SpecConstantComposite` intrinsic calls produced by the
`SpecConstants` pass:
```

%struct.A = type { i32, float }

; Function Attrs: alwaysinline
define dso_local spir_func void @get(%struct.A* sret %ret.ptr) local_unnamed_addr #0 {
  ; args are "ID" and "default value":
  %1 = tail call spir_func i32 @_Z20__spirv_SpecConstantii(i32 42, i32 0)

>   %2 = tail call spir_func i32 @_Z20__spirv_SpecConstantif(i32 43, float 0.000000e+00)

Reviewer 2: ```suggestion
  %2 = tail call spir_func float @_Z20__spirv_SpecConstantif(i32 43, float 0.000000e+00)
```

Author: Thanks for the catch, fixed in 879566ac6f71c4a82d1bf2eb7933df584982b90c

  %ret = tail call spir_func %struct.A @_Z29__spirv_SpecConstantCompositeif(%1, %2)
  store %struct.A %ret, %struct.A* %ret.ptr
  ret void
}
```

the translator will generate `OpSpecConstant` and `OpSpecConstantComposite`
SPIR-V instructions with proper `SpecId` decorations:

```
              OpDecorate %i32 SpecId 42                        ; ID
              OpDecorate %float SpecId 43                      ; ID
       %i32 = OpSpecConstant %int.type 0                       ; Default value
     %float = OpSpecConstant %float.type 0.0                   ; Default value

>     %struct = OpSpecConstantComposite %struct.type %i32 %float ; No ID, defined by its elements

Reviewer 2: The comment might be confusing.

Author: Tried to add more descriptive comments in 879566ac6f71c4a82d1bf2eb7933df584982b90c

         %1 = OpTypeFunction %struct.type

       %get = OpFunction %struct.type None %1
         %2 = OpLabel
              OpReturnValue %struct
              OpFunctionEnd
```

##### The post-link tool changes

For composite specialization constants the post link tool will additionally

> generate linearized list of \<leaf spec ID,offset,size\> tuples (descriptors),

Reviewer 2: I think it's worth explaining what is `offset` and why we have it here.

Author: Added a description in 879566ac6f71c4a82d1bf2eb7933df584982b90c

 - The SpecConstants pass in the post-link will have the following IR as input (`sret` conversion is omitted for clarity):

```
%struct.POD = type { [2 x %struct.A], i32 }
%struct.A = type { i32, float }

%spec_const = call %struct.POD __sycl_getSpecConstantValue<POD type mangling> ("MyConst_mangled")
```

> Based on the fact that `__sycl_getSpecConstantValue` returns `llvm::StructType`,

Reviewer 3: handling of `__sycl_getCompositeSpecConstantValue` is significantly different from handling of `__sycl_getSpecConstantValue`, that's why I designed it this way. I don't think overloading brings extra clarity here

Author: Returned this intrinsic back in e40b404e10c2bad9b4b40bf501ff6cd2bb17b2f4

it will be replaced with a set of `__spirv_SpecConstant` calls for each member
of its return type plus one `__spirv_SpecConstantComposite` to gather members
back into a single composite. If any composite member is another composite, then
it will be also represented by number of `__spirv_SpecConstant` plus one
`__spirv_SpecConstantComposite`:

```
%gold_POD_A0_x = call i32 __spirv_SpecConstant(i32 10, i32 0)
%gold_POD_A0_y = call float __spirv_SpecConstant(i32 11, float 0)

%gold_POD_A0 = call %struct.A __spirv_SpecConstantComposite(i32 %gold_POD_A0_x, float %gold_POD_A0_y)

%gold_POD_A1_x = call i32 __spirv_SpecConstant(i32 12, i32 0)
%gold_POD_A1_y = call float __spirv_SpecConstant(i32 13, float 0)

%gold_POD_A1 = call %struct.A __spirv_SpecConstantComposite(i32 %gold_POD_A1_x, float %gold_POD_A1_y)

%gold_POD_A = call [2 x %struct.A] __spirv_SpecConstantComposite(%struct.A %gold_POD_A0, %struct.A %gold_POD_A1)
%gold_POD_b = call i32 __spirv_SpecConstant(i32 14, i32 0)

> %gold = call %struct.POD __spirv_SpecConstantComposite([2 x %struct.A] %gold_POD_A, i32 %gold_POD_b)

Reviewer 3: the only reason why `__spirvCompositeSpecConstant` accepted IDs instead of SSA values is a concern that there could be some non-trivia data flow (i.e. via store/load to alloca) of an operand, which might confuse the translator. If we can be sure that this will never happen, then using SSA with spec constant tree construction in IR seems fine.

Author: Returned the intrinsic back in e40b404e10c2bad9b4b40bf501ff6cd2bb17b2f4

Spec ID for the composite spec constant is not needed, as runtime will never use it - it will use IDs of the leaves instead.
 Yet, the SPIR-V specification does not allow `SpecID` decoration for composite spec constants.
Spec ID for the composite spec constant is not needed, as runtime will never use
it - it will use IDs of the leaves instead.
Yet, the SPIR-V specification does not allow `SpecID` decoration for composite
spec constants, because its defined by its members instead.

##### The post-link tool changes

For composite specialization constants the post link tool will additionally
generate linearized list of \<leaf spec ID,type,offset,size\> tuples (descriptors),
generate linearized list of \<leaf spec ID,offset,size\> tuples (descriptors),
where each tuple describes a leaf field, and store it together with the
existing meta-information associated with the specialization constants and
passed to the runtime. Also, for a composite specialization constant there is
no ID map entry within the meta information, and the composite constant is
referenced by its symbolic ID. For example:

```
MyConst_mangled [10,int,0,4],[11,float,4,4],[12,int,8,4],[13,float,12,4],[14,int,16,4]
```

> #### LLVMIR-\>SPIR-V translator

Reviewer 3: Why is this section removed? Are you proposing not to change LLVM translator at all?

Author: > Why is this section removed?

From my point of view it is not really useful, as interface between LLVM IR and SPIR-V is owned by the translator and not by intel/llvm. We use this interface (SPIR-V friendly IR) for almost every SYCL built-in and we don't have an extensive documentation for the format for each built-in, because we just refer to a document in the upstream translator repo.

> Are you proposing not to change LLVM translator at all?

It still needs to be changed, but I don't see anything special which needs to be documented here: I would rather update [SPIR-V friendly IR document](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/docs/SPIRVRepresentationInLLVM.rst) to clarify the exact format if that is needed.

Reviewer 3: I do think that the design doc would benefit a lot from at least high-level description of what happens in the translator, as this is part of compilation process, even though separated into a different component.

Author: > I do think that the design doc would benefit a lot from at least high-level description of what happens in the translator, as this is part of compilation process, even though separated into a different component.

Ok, makes sense. I will leave some high-level description of the interface with the SPIR-V translator then

Author: Returned this section back in e40b404e10c2bad9b4b40bf501ff6cd2bb17b2f4
